servlet/src/jettyTest/java/io/grpc/servlet/JettyTransportTest.java

```diff
@@ -81,7 +88,25 @@ public void start(ServerListener listener) throws IOException {
         httpConfiguration.setDelayDispatchUntilContent(false);
 
         HTTP2CServerConnectionFactory factory =
-                new HTTP2CServerConnectionFactory(httpConfiguration);
+                new HTTP2CServerConnectionFactory(httpConfiguration) {
+                  @Override
+                  protected ServerSessionListener newSessionListener(Connector connector,
+                                                                     EndPoint endPoint) {
+                    return new HTTPServerSessionListener(connector, endPoint) {
+                      @Override
+                      public void onSettings(Session session, SettingsFrame frame) {
+                        HTTP2Session s = (HTTP2Session) session;
+                        // Override client's requested SETTINGS so
```

Comment:
I thought the test would allow either the client or the server to fail the request because of too-large headers. I know for Netty the server enforces the header size when sending, and I thought that test passes.

Reply:
The test seems to only confirm that the client receives UNKNOWN, RESOURCE_EXHAUSTED, or INTERNAL, but when jetty validates that the header being sent was too big, it seems to kill the whole transport (not just the stream). That results in the netty client emitting to the test the UNAVAILABLE grpc status, which fails the assertion.

With the workaround in place, the value of `status.getCode()` is `INTERNAL`, which appears to come from the netty client. The netty client returns the same status code when talking to a netty server.

Jetty is responding with 200 OK, the too-large metadata (but only with this workaround to let it), and then seems to contain the rest of the expected response? Netty on the other hand sends only a RST_STREAM.

Tomcat and Undertow send the 200, too-large metadata, and the payload. Like Jetty, the client then handles the error, by noticing that the header was too large, and replacing the OK grpc status with INTERNAL.

It seems like this test points out that the netty client handles a RST_STREAM as "INTERNAL", but FIN-ing the whole http stream is "UNAVAILABLE". 

The test is apparently about the _client_ checking the inbound header sizes, but from these observations, it seems that when run against a netty server, the client can't actually check the header size, because the header is never sent?

Reply:
> jetty validates that the header being sent was too big, it seems to kill the whole transport (not just the stream)

Wow. That's really broken for _sending_ direction. In the past, we've filed bugs about such things and then just disabled the impacted tests. UNAVAILABLE is wrong, as the client will retry and the server will produce the same response! Seems like there are multiple bugs there (they shoudn't kill the connection, and if they kill the connection they need to use a proper GOAWAY error code).

> The test is apparently about the client checking the inbound header sizes

Yeah, in HTTP/2 only the receiver needs to check the size. But some senders started checking due to misreading the spec, and with many receivers getting pretty strict with CONTINUATIONS recently more senders are probably checking.

Reply:
If GOAWAY is expected, netty is also wrong, sending a RST_STREAM instead - GOAWAY kills the connection itself right?

The UNAVAILABLE is not coming from Jetty, but netty-client's own interpretation of what the killed connection means (again, in contrast to netty server, ending with GOAWAY, which signifies INTERNAL).

I can file an issue with Jetty to ask them to change their implementation? I'm not seeing a specific note in the spec about this - do you have a reference you could point me at that I could understand better and discuss with them? Specifically, I looked at `SETTINGS_MAX_HEADER_LIST_SIZE`, but the only specific note I see is (10.5.1):

> This setting is only advisory, so endpoints MAY choose to send header blocks that exceed this limit and risk having the request or response being treated as malformed. This setting is specific to a connection, so any request or response could encounter a hop with a lower, unknown limit.

5.4.1. seems relevant too, but doesn't require (only encourage) that a server send GOAWAY first:

> An endpoint can end a connection at any time. In particular, an endpoint MAY choose to treat a stream error as a connection error. Endpoints SHOULD send a [GOAWAY](https://httpwg.org/specs/rfc7540.html#GOAWAY) frame when ending a connection, providing that circumstances permit it.

Reply:
GOAWAY is not expected, per se. The expectation is the stream would get killed. But if an implementation decides to kill the connection, it should send GOAWAY with an error code (which you quoted).

> The UNAVAILABLE is not coming from Jetty, but netty-client's own interpretation of what the killed connection means

The killing of the connection is the issue. With wireshark, yeah, I see Jetty just hang up because the server tried to send too much. That's just silly. Let's not work around that; it's better to disable the test.

[jetty killed connection.pcap.gz](https://github.com/grpc/grpc-java/files/15341243/jetty.killed.connection.pcap.gz)

Reply:
Sounds good. Yeah, with most of this stuff we don't really expect it to be backported.

When you add the `@Ignore`, having a comment on that line with a link to the Jetty bug would be fabulous.

> grpc-java's gradle version apparently can't handle Java 17

Our CI runs with Java 17. I just did a build with Java 21. I think Java 22 might not work.

Reply:
Ah sorry - I was referring to the "what java version is running" check in servlet/jakarta/build.gradle that isn't able to test for 17, to make running the test conditional on the running version.

Reply:
Ah, I think you can do `JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_17)` (untested!)

Reply:
That does work, thanks - when this issue is resolved, I'll see about adding Jetty 12 into the mix guarded by that check. It will also need at least those two tests un-ignored, and slightly different setup code, so will probably warrant its own sourceSet.

Reply:
You could maybe turn off the tests for the older Jetty versions at that point. Something for the future, though.